### DISCUSSION TITLE
Retrieving a Function Result in C++

### DISCUSSION THREAD

**Discussion by dhathris** [casadi/casadi#3782] [Q&A]
Hello, I am trying to understand how to retrieve a function call results. I have the following function definition:
`Function derivative = Function("derivative", {mx_gip, mx_X}, {mx_der}, {"t", "sX"}, {"derivative"}, opts);`
with the following function call:
`derivative.call(args, der_dm);`
Now is the expected data type of `der_dm` output variable of type `std::vector<DM>`? If so how can I retrieve the values from this variable, `der_dm`? Assuming vector retrieval function call such as `der_dm.at(ii)` is not working. Note that the output of the function `mx_der` is defined as an MX variable.

**Comment by tmmsartor** [comment]
I think you are defining datatype of `der_dm` multiple options are possible, i.e. `std::vector<DM>` or `DMDict`. It is important to notice that input and output may have nested indices, in your case you have 2 input and 1 output, then each of those element can be a vector (or matrix) itself. 
To access the numeric values, first you have to select the right output, in your case only `der_dm.at(0)` will work, and then the values inside the that first output.
I attach an example to hopefully clarify:

```
#include <casadi/casadi.hpp>
#include <iostream>

using namespace casadi;
using namespace std;

int main() {
  MX x = MX::sym("x");
  MX y = MX::sym("y");
  MX in = vertcat(x,y);
  MX out = vertcat(pow(x,2)/y,x);
  Function f = Function("f",{in},{out},{"in"},{"out"});

  DMDict din,dout;
  din["in"] = {2.0,3.0};
  dout["out"] = {0.0, 0.0};
  f.call(din, dout);

  DM dmres = dout.at("out");
  auto res = vector<double>(dmres);
  cout << res.at(0) << ", " << res.at(1) << endl;

  // equivalent
  
  vector<DM> dmin {{2.0,3.0}};
  vector<DM> dmout {{0.0, 0.0}};
  f.call(dmin, dmout);
  DM dmres2 = dmout.at(0);
  auto res2 = vector<double>(dmres2);
  cout << res2.at(0) << ", " << res2.at(1) << endl;

}
```
